runtime/src/mm/cpp/Memory.cpp

```diff
@@ -214,19 +274,37 @@ extern "C" RUNTIME_NOTHROW OBJ_GETTER(AdoptStablePointer, void* pointer) {
     auto* threadData = mm::ThreadRegistry::Instance().CurrentThreadData();
     auto* node = static_cast<mm::StableRefRegistry::Node*>(pointer);
     ObjHeader* object = **node;
-    UpdateReturnRef(OBJ_RESULT, object);
+    // Make sure `object` stays in the rootset: put it on the stack before removing it from `StableRefRegistry`.
+    mm::SetStackRef(OBJ_RESULT, object);
     mm::StableRefRegistry::Instance().UnregisterStableRef(threadData, node);
     return object;
 }
 
 extern "C" RUNTIME_NOTHROW void CheckLifetimesConstraint(ObjHeader* obj, ObjHeader* pointee) {
+    // TODO: Consider making it a `RuntimeCheck`. Probably all `RuntimeCheck`s and `RuntimeAssert`s should specify
+    //       that their firing is a compiler bug and should be reported.
     if (!obj->local() && pointee != nullptr && pointee->local()) {
         konan::consolePrintf("Attempt to store a stack object %p into a heap object %p\n", pointee, obj);
         konan::consolePrintf("This is a compiler bug, please report it to https://kotl.in/issue\n");
         konan::abort();
     }
 }
 
+extern "C" ALWAYS_INLINE bool TryAddHeapRef(const ObjHeader* object) {
+    // TODO: Remove when legacy MM is gone.
+    return true;
```

Comment:
I'm not sure this implementation is correct. OTOH, I guess we can't provide a proper implementation at the moment. Can we leave it with `TODO`?